날씨정보(전) [Step3]  yescoach

YesCoach opened the pull request:
웨더 안녕하세요!! @SungPyo 
Step3 구현 해보고 이렇게 PR을 보냅니다🙂🙂
이번 스텝도 잘 부탁드립니다!!! 🙏🙏

## 고민한점
- ViewController와 datasource의 분리에 대해서 리뷰어님께서 해주신 내용, 동료 캠퍼가 이야기 해주신 내용 참고해서 생각해보았는데요,
ViewController에서 tableView DataSource를 분리하는 것은 DataSource를 분리 구현함으로써 다른 DataSource가 얼마든지 사용될 수 있다는것이고,, 테이블뷰가 상황에 따라 다른 셀 구성이 필요할때 DataSource의 분리가 의미있을 것 같아요.
저번 LocationManager도 그렇고 분리했을때 두 객체간에 통신? 참조? 하는 방법으로 Delegate와 클로저를 말씀해주셨는데 아직은 어떤 방향인지 감이 안와서(특히 클로저를 활용하는것이 어떤 형태인지 모르겠어요. 클로저에 아직 익숙치가 않아서...) 고민이 길어질 것 같아 기존처럼 뷰컨에 extension으로 datasource와 delegate를 구현했습니다. 이번 날씨정보 프로젝트에서는 상황에 따라 tableView의 datasource를 변경할 일이 없기 때문에 꼭 datasource를 분리해야 되는 건 아닌 것 같기도 하구요...ㅠㅠ

## 궁금한 점
- 현재 기기의 Locale을 받기 위해서 `Locale.current`를 사용하려고 했는데요!! 시뮬레이터가 이상한건지 설정이 `한국어 - 한국` 으로 되어 있음에도 `en_KR` 값으로 반환되더라구요... 그래서 일단은 preferredLanguage의 first값으로 Locale을 생성하여 사용하고 있습니다. Locale.current를 했을때 반환되는 Locale 중 언어와 관련된 값은 현재 기기에 설정되어 있는 주 사용 언어가 값으로 전달되는게 맞을까요??
- 테이블뷰에 separator가 있는데 Cell에만 적용하고 헤더뷰에는 적용시키지 않는 방법이 있을까요? CALayer를 활용해서 separator를 가려주는 내용이 있었던 것 같은데 다른 방법이 또 있을지 궁금합니다!!
- 
``` swift
2021-10-14 13:12:07.698339+0900 WeatherForecast[3200:105156] [boringssl] boringssl_metrics_log_metric_block_invoke(144) Failed to log metrics
```
프로젝트를 실행시키면 이런 에러 메세지가 등장하는데... 어떤 부분에서 에러를 발생시키는 건지 모르겠습니다ㅠㅠ 검색해보니 network 통신과 관련이 있는 것 같은데 어떤 에러인지 궁금합니다ㅠㅠ
- 마지막으로 UI를 구현할때 항상 기준이 없어서 고민이 많이 됩니다. 어떤 분은 연산 프로퍼티로 UI 객체를 구현하시기도 하고... 저처럼 프로퍼티를 일단 생성하고 메소드를 만들어 거기서 구체적인 설정을 하기도 하고... 하물며 뷰에 `addSubview` 해주는 것과 `translatesAutoresizingMaskIntoConstraints`의 값에 false로 주는 것, constraints을 지정해주는 것의 순서나 위치같은 것도 항상 달라지는 것 같아서요!! 이런것 또한 스타일이라면 리뷰어님께서는 어떤 기준을 갖고 계신지 궁금합니다!!

hrjy6278 committed:
docs: README.md 수정

hrjy6278 committed:
docs: README.md 오탈자 수정

YesCoach committed:
docs: README.md 스타일 추가

details 및 목차 링크 추가했습니다.

YesCoach committed:
Merge branch 'step2' of https://github.com/YesCoach/ios-weather-forecast into step3

YesCoach committed:
refactor: Fetching Model 중 사용하지 않는 값 삭제

YesCoach committed:
refactor: LocationManagerDelegate 수정

- LocationManagerDelegate의 메소드 불필요한 매개변수 삭제
- ViewController에서 패칭한 후 프로퍼티에 할당하도록 구현

YesCoach committed:
feat: CustomTableViewCell 및 dataSource 생성

YesCoach committed:
feat: ViewController내 TableView 구현

YesCoach committed:
refactor: ViewController 내 dataSource 구현

- WeatherForecastViewController에 UITableViewDataSource 채택 및 구현
- WeatherForecastViewCell의 CellConfigure 매소드 구현

YesCoach committed:
feat: TableViewCell Configure 상세구현

- 날짜를 형식에 맞게 변환하기 위해 DateFormatter 적용
- 온도를 Kelvin에서 Celsius로 변환하기 위해 MeasurementFormatter 적용
- 날씨 아이콘을 URL로 부터 받아올때 캐시를 사용하기 위한 로직 구현
- data를 받아오면 tableView를 비동기적으로 다시 그리기 위한 적용

YesCoach committed:
feat: VC에 tableDelegate 구현 및 Header 구현

- WeatherForecastViewController에 UITableViewDelegate 채택
- TableView에 Header 구현
- Header configure을 위해 fetchingWeatherData 로직 수정
- TableView Cell 높이 지정

YesCoach committed:
refactor: 요구 format에 맞게 Cell contents 수정

- measurementFormatter 온도 변환 메소드 반환값 양식 수정
- CLGeocoder 주소 반환 메소드 매개변수 수정(Locale-"ko_KR")
- WeatherForecastHeaderView 레이블 포맷 수정

YesCoach committed:
refactor: 배경 추가, font 흰색으로 변경, Cell 배경 투명화

- UILabel extension, 폰트 색을 지정할 수 있는 convenience init 구현
- 배경 이미지가 보이도록 Cell의 backgroundColor를 .clear로 설정

YesCoach committed:
feat: UIRefreshControl 기능 구현

- UITableView의 프로퍼티 refreshControl에 UIRefreshControl 객체 추가
- Refreshing시 requestLocation 수행하고 tableView를 새로고침
- 새로고침과 동시에 Refreshing 종료

YesCoach committed:
refactor: 접근제어 설정, 불필요한 코드 제거, 하드코딩 수정

- 사용하지 않는 WeatherForecastViewDataSource 제거
- 불필요한 tableView.layoutIfNeeded 메소드 사용 코드 제거
- 가능한 메소드 모두 private로 접근제어 설정
- ImageURL의 하드코딩을 피하기 위해 WeatherAPI 타입에 static 변수 생성

YesCoach committed:
refactor: 현재 기기 설정의 Locale을 받아오도록 코드 수정

- DateFormatter에서 Locale을 설정할때 기기의 Locale을 받아오도록 로직 구현

SungPyo commented on WeatherForecast/WeatherForecast/Controller/WeatherForecastViewController.swift:
```diff
@@ -0,0 +1,133 @@
+// MARK: UITableViewDataSource 구현부
+extension WeatherForecastViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return forecastData?.list.count ?? 0
```
스타일이지만 현재 프로젝트만 봤을때 요러한 옵셔널 처리를 간편하게 하기 위해 forecastData 를 ? 가 아닌 = [] 로 초기화 하는걸 조금 더 추천합니다! CellforRow에도 비슷한게 있네요 ㅎㅎ

YesCoach replied:
해결했습니다!! 이전에 오픈마켓때도 이렇게 썼던 것 같은데 이런 처리가 훨씬 나아 보이네요😀😀

SungPyo commented on WeatherForecast/WeatherForecast/Controller/WeatherForecastViewController.swift:
```diff
@@ -0,0 +1,133 @@
+                }
+            }
+            DispatchQueue.main.async {
+                self.tableView.reloadData()
```
이 함수는 데이터를 fetch하는 부분이므로 tableView의 reload는 해당 함수에는 어울리지 않다고 생각되는데 예스코치의 다른 의도가 있었을까요?!

YesCoach replied:
여기서는 데이터 fetch가 끝난 시점에 tableView를 reload하고자 이 위치에 넣어줬습니다! 다만 구현할때도 아닌것 같았는데 리뷰어님 말씀을 듣고 든 생각으로, data fetching이라는 함수의 역할과 tableView의 reload는 맞지 않는 것 같습니다. 패칭을 하는 메소드는 패칭하는 동작 그대로 두고, 테이블 뷰를 데이터에 맞게 다시 그리는 로직은 didUpdateLocation() 이라던지 다른 곳에서 수행하는게 더 자연스럽겠죠??
이 경우 데이터 패칭이 끝난 시점에 reload를 하려면 dispatch group을 사용하는게 좋을 것 같은데 한번 적용해서 리팩토링 진행해보도록 하겠습니다.

SungPyo commented on WeatherForecast/WeatherForecast/Controller/WeatherForecastViewController.swift:
```diff
@@ -0,0 +1,133 @@
+// MARK: UITableViewDelegate
+extension WeatherForecastViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return view.frame.height / 15
```
이건 그냥 질문인데요! 가변 높이에 대한 estimatedHeight 에 대한 기능은 잘 알고 계신가요?!

YesCoach replied:
어설프게 알고 있었는데 이번 기회에 다시 정확히 공부해보려고 합니다!
estimatedHeight는 테이블뷰 인스턴스의 프로퍼티로 셀 높이의 추정치를 전달합니다. 셀의 contents에 따라 높이가 가변적으로 설정되는 경우에, 모든 행의 높이를 계산하는 것은 그만큼 높은 비용이 들기 때문에 estimatedHeight를 줌으로써 load와 scroll 시 드는 비용을 줄일 수 있을 것 같습니다.

이번 프로젝트에서는 셀마다 띄워주는 콘텐츠의 크기가 일정하다고 생각되어서 고정 높이를 `heightForRowAt` 메소드로 지정해 주었습니다!
또 한가지는... 테이블뷰의 프로퍼티 중에 rowHeight가 있는데...
헉 지금 확인했네요!! 공식문서에서 테이블뷰의 셀 높이가 고정이라면 이 메소드를 쓰지 말고 rowHeight에 지정하라고 하네요...
이 메소드는 고정 높이가 아닌 경우 사용하라고 하는군요... 적으면서 배웠습니다! 다시 정리해보겠습니다!

SungPyo commented on WeatherForecast/WeatherForecast/Controller/WeatherForecastViewCell.swift:
```diff
@@ -0,0 +1,69 @@
+
+        if let iconID = data.weather.first?.icon {
+            let iconURL = WeatherAPI.imagebaseURL + iconID + ".png"
+            iconImageView.setImageURL(iconURL)
```
prepareForReuse 에서 초기화 이후에 이미지 요청에 대한 응답이 오는경우가 발생할 수 있는데, 해당 상황에 경우 어떻게 처리해야할지 한번만 고민해주세요!

YesCoach replied:
말씀해주신 덕분에 발생 가능한 문제를 인지할 수 있었습니다!! image를 받아올때 처음 받는 이미지면 cache에 넣는 작업을 하고, 이때 url을 cacheKey로 설정하고 넣는데요! data를 받은 다음 메인 스레드에서 이미지 넣을때 컴플리션 핸들러를 사용해서 해당 셀의 url과 cacheKey로 비교하고, 두 값이 같을때만 넣어주도록 하는건 어떨지 생각해보았습니다...

SungPyo reviewed (approved):
예스코치 안녕하세요!

바쁜일이 있어 리뷰가 늦었어요 죄송합니다!

고민하신 내용은 아마 이후에도 계속계속 고민이 될 것이고, 불편함을 느낀 순간 결정하게 되실거기때문에 이러한 고민을 했다. 정도로 기억하시면 충분할거라 생각됩니다 ㅎㅎ

1. 현재 기기의 Locale을 받기 위해서 Locale.current를 사용하려고 했는데요!! 시뮬레이터가 이상한건지 설정이 한국어 - 한국 으로 되어 있음에도 en_KR 값으로 반환되더라구요... 그래서 일단은 preferredLanguage의 first값으로 Locale을 생성하여 사용하고 있습니다. Locale.current를 했을때 반환되는 Locale 중 언어와 관련된 값은 현재 기기에 설정되어 있는 주 사용 언어가 값으로 전달되는게 맞을까요??
ㄴ`저도 같은 이유로 인해 preferredLanguage.first 하고 있습니닷. 자세한 내용은 제가 직접 써드리는 것 보다 > https://stackoverflow.com/questions/48136456/locale-current-reporting-wrong-language-on-device <  를 한번 보시는것도 좋을 것 같아요!`

2. 테이블뷰에 separator가 있는데 Cell에만 적용하고 헤더뷰에는 적용시키지 않는 방법이 있을까요? CALayer를 활용해서 separator를 가려주는 내용이 있었던 것 같은데 다른 방법이 또 있을지 궁금합니다!!
ㄴ`요 부분은 실무에서 디자이너별로 많이 달라서 저의 경우 기본 separator를 없애고 해당 UI 디자인에 추가하는 방식을 사용합니다. 하지만 꼭 정해진 방법이 있는것은 아니니 디자인에 따라서 유동적으로 대응할 수 있는 어떠한 방식이든 무관할 것 같아요.`

3. 프로젝트를 실행시키면 이런 에러 메세지가 등장하는데... 어떤 부분에서 에러를 발생시키는 건지 모르겠습니다ㅠㅠ 검색해보니 network 통신과 관련이 있는 것 같은데 어떤 에러인지 궁금합니다ㅠㅠ
ㄴ`해당 에러는 iOS에서 뱉어주는 에러로 말씀하신 대로 network 에러같은데 데이터가 올바르게 로드된다면 문제가 없다고 판단됩니다! 통신에 문제를 일으킬만한 코드가 보이지는 않는 것 같은데...! 저라면 '일단' 무시할 것 같아요!`

4. 마지막으로 UI를 구현할때 항상 기준이 없어서 고민이 많이 됩니다. 어떤 분은 연산 프로퍼티로 UI 객체를 구현하시기도 하고... 저처럼 프로퍼티를 일단 생성하고 메소드를 만들어 거기서 구체적인 설정을 하기도 하고... 하물며 뷰에 addSubview 해주는 것과 translatesAutoresizingMaskIntoConstraints의 값에 false로 주는 것, constraints을 지정해주는 것의 순서나 위치같은 것도 항상 달라지는 것 같아서요!! 이런것 또한 스타일이라면 리뷰어님께서는 어떤 기준을 갖고 계신지 궁금합니다!!
ㄴ`우선 코딩 스타일이라고 생각됩니다. 각 회사별로 스타일이 있을것이고, 저의 경우는 이러한 스타일이 크게 문제가 되지 않는다고 생각하기에 따르는 편이구요! 개인적인 스타일로는 초기화 후 추가 함수에서 뷰를 그려주는 방식을 선호합니다 ㅎㅎ`

예스코치 고생하셨어요! 문제가 전혀 없다고 할 수는 없지만 따로 코멘트를 달아드렸으니 해당 부분만 다음 스텝 진행하실때 확인하시면 될 것 같아요!

SungPyo commented on WeatherForecast/WeatherForecast/Controller/WeatherForecastViewController.swift:
```diff
@@ -0,0 +1,133 @@
+// MARK: LocationManagerDelegate 구현부
+extension WeatherForecastViewController: LocationManagerDelegate {
+    func didUpdateLocation() {
+        fetchingWeatherData(api: WeatherAPI.current, type: CurrentWeather.self)
+        fetchingWeatherData(api: WeatherAPI.forecast, type: ForecastWeather.self)
```
요 부분이 어떤 문제를 일으킬 것 같은지 의견을 들어보고 싶습니다!

YesCoach replied:
음... 일단 해당 메소드는 패칭하는 동작을 하기 때문에 비동기로 동작하고... 데이터를 받았을때 테이블뷰를 다시 그리는 동작이 메소드 내부에 있기 때문에 총 두번 다시 그리게 되는 것 까지는 짐작됩니다ㅜㅜ reloadData는 패칭이 다 끝난 시점에 동작하도록 리팩토링 하겠습니다. 혹시 다른 문제도 있을까요??

SungPyo replied:
한가지 이상의 문제가 있다면 또다른 문제가 있을'수' 도 있는거라서 문제가 없다고 판단되었을떄 다시 보는게 좋을 것 같아요!

YesCoach committed:
refactor: 이미지 지연로딩 대응하는 로직 구현

cacheKey와 image url을 비교하여 다른 경우 이미지를 넣지 않는 로직 추가

YesCoach committed:
refactor: DataFetching과 tableView.reloadData 분리

- 데이터를 패칭하는 메소드로부터 tableView를 다시 그리는 reloadData 동작을 분리구현
- 데이터를 모두 패칭 완료했을때 뷰를 그리기 위해 dispatchGroup으로 완료 시점 확인
- 헤더뷰에 적용하지 못했던 이미지 지연로딩 코드 추가

YesCoach committed:
refactor: ViewController의 forecastData 타입 수정
- 옵셔널을 피하기 위해 [ForecastWeather.List] = [] 구현
